js/ui/control/geolocate_control.js

```diff
@@ -104,13 +112,30 @@ class GeolocateControl extends Evented {
     }
 
     _onClickGeolocate() {
-
-        window.navigator.geolocation.getCurrentPosition(
-            this._onSuccess, this._onError, geoOptions);
-
-        // This timeout ensures that we still call finish() even if
-        // the user declines to share their location in Firefox
-        this._timeoutId = setTimeout(this._finish, 10000 /* 10sec */);
+        const positionOptions = util.extend(defaultGeoPositionOptions, this.options && this.options.positionOptions || {});
+
+        // toggle watching the device location
+        if (this.options.watchPosition) {
+            if (this._watching) {
```

Comment:
Would it be possible to eliminate the `_watching` flag and use `_geolocationWatchID` as the single source of truth about which state we're in?